tests/lean/run/discrTreeNumeral.lean

```diff
@@ -0,0 +1,144 @@
+example : MyProp (OfNat.ofNat (n + 1)) := by
+  dsimp only [myProp_ofNat_add_one]
+
+/-- Lemmas about numerals that express a lower-bound
+  through a constant offset match a larger offset.-/
+example : MyProp (OfNat.ofNat (n + 2)) := by
+  dsimp only [myProp_ofNat_add_one]
+
+/-- Lemmas about numerals that express a lower-bound
+  through a constant offset match the constant offset.-/
+example : MyProp 1 := by
+  dsimp only [myProp_ofNat_add_one]
+
+/-- Lemmas about numerals that express a lower-bound
+  through a constant offset match numerals exceeding the constant offset.-/
+example : MyProp 2 := by
+  dsimp only [myProp_ofNat_add_one]
+
+/-- Lemmas about `0` match `Nat.zero` since `Nat.zero = 0` is in the default simp set.-/
+example : MyProp Nat.zero := by
+  dsimp [myProp_zero]
+
+/-- Lemmas about `0` match `nat_lit 0` since `nat_lit 0 = 0` is in the default simp set.-/
+example : MyProp (nat_lit 0) := by
+  dsimp [myProp_zero]
+
+/-- Lemmas about any natural number plus a constant offset match their LHS in terms of `Nat.succ`
+  since `Nat.succ n = n + 1` is in the default simp set. -/
+example : MyProp (Nat.succ n) := by
+  dsimp [myProp_add_one]
+
+/-- If a general lemma about numerals is specialized to a particular one, it still applies.
+
+  This is a tricky case because it results in nested `OfNat.ofNat`.
+  For example, `(myProp_ofNat 0 : MyProp (OfNat.ofNat (OfNat.ofNat (nat_lit 0)))`.
+
+  This currently works because we're passing an expression instead of a declaration name,
+  so `simp` considers it a local hypothesis and bypasses precise discrimination tree matching.
+  -/
+example : MyProp 0 := by
+  dsimp only [myProp_ofNat 0]
+
+/-- If a general lemma about natural numbers plus a constant offset is specialized
+  to a constant base, it still applies.
+
+  This is a tricky case because it results in a non-simp-normal lemma statement.
+  For example, `(myProp_add_one 0 : MyProp (0 + 1))` can be simplified to `MyProp 1`.-/
+example : MyProp 1 := by
+  dsimp only [myProp_add_one 0]
+
+/-- If a general lemma about numerals that expresses a lower-bound
+  through a constant offset is specialized to a constant base, it still applies.-/
+example : MyProp 1 := by
+  dsimp only [myProp_ofNat_add_one 0]
+
+class MyClass (n : Nat) : Prop where
+
+instance myClass_succ [MyClass n] : MyClass (n.succ) := ⟨⟩
+
+section BaseZero
+
+local instance myClass_nat_zero : MyClass (Nat.zero) := ⟨⟩
+
+/-- A base instance for `P Nat.zero` and a step instance for `P n → P (n.succ)` can combine to reach any
+  sequence of `Nat.zero...succ.succ`.
+
+  Because the kernel has special handling for unifying `Nat.succ`,
+  this requires that they can also combine to reach any numeral.
+
+  For example, to reach `P (Nat.zero.succ.succ)`,
+  the first recursive step contains the unification problem `P (Nat.zero.succ.succ) =?= P (?n.succ)`,
+  which the kernel solves by setting `?n := 1`, *not* `?n := Nat.zero.succ`.
```

Comment:
It just occurred to me to test this for `nat_lit 0`, and lo and behold, unification has the same surprising behavior of setting the metavariable to `1`:

```lean
set_option pp.natLit true in
#check (show (nat_lit 0).succ.succ = Nat.succ _ from rfl)
-- Nat.succ (Nat.succ (nat_lit 0)) = Nat.succ 1
```

If we replace `Nat.zero` with `nat_lit 0` in this test case, we get a new test case which fails after this PR:
```lean
class MyClass (n : Nat) : Prop where

instance myClass_succ [MyClass n] : MyClass (n.succ) := ⟨⟩

instance myClass_nat_lit_zero : MyClass (nat_lit 0) := ⟨⟩

example : MyClass (nat_lit 0).succ.succ :=
  inferInstance -- failed to synthesize instance
```

Making this instance path work as expected would require a significantly more complicated fix to the discrtree code (namely, normalize `nat_nit n` to `OfNat.ofNat (nat_lit n)`, except when already inside `OfNat.ofNat`), and the behavior after the current PR is easy to work around (add an instance for `0` instead of or in addition to `nat_lit 0`). Would it be acceptable to leave addressing this case as future work?